packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -1473,6 +1485,137 @@ class _SelectableFragment with Selectable, ChangeNotifier implements TextLayoutM
     return SelectionUtils.getResultBasedOnRect(_rect, localPosition);
   }
 
+  SelectionResult _updateSelectionEdgeByWord(Offset globalPosition, {required bool isEnd}) {
+    // When the start/end edges are swapped, i.e. the start is after the end, and
+    // the scrollable synthesizes an event for the opposite edge, this will potentially
+    // move the opposite edge outside of the origin word boundary and we are unable to recover.
+    final TextPosition? existingSelectionStart = _textSelectionStart;
+    final TextPosition? existingSelectionEnd = _textSelectionEnd;
+    _selectableContainsOriginWord ??= existingSelectionStart != null && existingSelectionEnd != null;
+
+    _setSelectionPosition(null, isEnd: isEnd);
+    final Matrix4 transform = paragraph.getTransformTo(null);
+    transform.invert();
+    final Offset localPosition = MatrixUtils.transformPoint(transform, globalPosition);
+    if (_rect.isEmpty) {
+      return SelectionUtils.getResultBasedOnRect(_rect, localPosition);
+    }
+    final Offset adjustedOffset = SelectionUtils.adjustDragOffset(
+      _rect,
+      localPosition,
+      direction: paragraph.textDirection,
+    );
+
+    final TextPosition position = paragraph.getPositionForOffset(adjustedOffset);
+    // Check if the original local position is within the rect, if it is not then
+    // we do not need to look up the word boundary for that position. This is to
+    // maintain a selectables selection collapsed at 0 when the local position is
+    // not located inside its rect.
+    final ({TextPosition wordStart, TextPosition wordEnd})? wordBoundary = !_rect.contains(localPosition) ? null : _getWordBoundaryAtPosition(position);
+    TextPosition? targetPosition;
+    if (wordBoundary != null) {
+      assert(wordBoundary.wordStart.offset >= range.start && wordBoundary.wordEnd.offset <= range.end);
+      if (existingSelectionStart != null && existingSelectionEnd != null) {
+        final TextPosition edgeToKeepInPlace = isEnd ? existingSelectionStart : existingSelectionEnd;
+        final TextPosition currentMovingEdge = isEnd ? existingSelectionEnd : existingSelectionStart;
+        // Swap the selection edges to keep the origin word in bounds, when the
+        // new position where to cause an inverted selection.
+        final bool shouldSwapEdgesWhenSelectionIsNormalized = isEnd ? position.offset < existingSelectionStart.offset : position.offset > existingSelectionEnd.offset;
+        final bool shouldSwapEdgesWhenSelectionNotNormalized = isEnd ? position.offset > existingSelectionStart.offset : position.offset < existingSelectionEnd.offset;
```

Comment:
This seems to be inverse of `shouldSwapEdgesWhenSelectionIsNormalized` except for the equal case? can the equal case be handle the same way as either one of `shouldSwapEdgesWhenSelectionIsNormalized` or `shouldSwapEdgesWhenSelectionNotNormalized`

Reply:
it is also hard to interpret the name.  is it swaping the `existingSelectionStart` and `existingSelectionEnd` or is it swaping the targetPosition to be `wordBoundary.wordStart` or `wordBoundary.wordEnd`? base on the code, it seems to be both.

Maybe call this `willNewSelectionBeNormalized`

Reply:
The equal case should actually not swap, I have updated the implementation to reflect this.